client/fingerprint/vault.go

```diff
@@ -22,71 +23,90 @@ const (
 
 // VaultFingerprint is used to fingerprint for Vault
 type VaultFingerprint struct {
-	logger    log.Logger
-	client    *vapi.Client
-	lastState string
+	logger     log.Logger
+	clients    map[string]*vapi.Client
+	lastStates map[string]string
 }
 
 // NewVaultFingerprint is used to create a Vault fingerprint
 func NewVaultFingerprint(logger log.Logger) Fingerprint {
-	return &VaultFingerprint{logger: logger.Named("vault"), lastState: vaultUnavailable}
+	return &VaultFingerprint{
+		logger:     logger.Named("vault"),
+		clients:    map[string]*vapi.Client{},
+		lastStates: map[string]string{"default": vaultUnavailable},
+	}
 }
 
 func (f *VaultFingerprint) Fingerprint(req *FingerprintRequest, resp *FingerprintResponse) error {
-	config := req.Config
-
-	if config.VaultConfig == nil || !config.VaultConfig.IsEnabled() {
-		return nil
+	for _, cfg := range f.vaultConfigs(req) {
+		err := f.fingerprintImpl(cfg, resp)
+		if err != nil {
+			return err
+		}
 	}
+	return nil
+}
+
+// fingerprintImpl fingerprints for a single Vault cluster
+func (f *VaultFingerprint) fingerprintImpl(cfg *config.VaultConfig, resp *FingerprintResponse) error {
 
 	// Only create the client once to avoid creating too many connections to Vault
-	if f.client == nil {
-		vaultConfig, err := config.VaultConfig.ApiConfig()
+	client := f.clients[cfg.Name]
+	if client == nil {
+		vaultConfig, err := cfg.ApiConfig()
 		if err != nil {
-			return fmt.Errorf("Failed to initialize the Vault client config: %v", err)
+			return fmt.Errorf("Failed to initialize the Vault client config for %s: %v", cfg.Name, err)
 		}
-		f.client, err = vapi.NewClient(vaultConfig)
+		client, err = vapi.NewClient(vaultConfig)
 		if err != nil {
-			return fmt.Errorf("Failed to initialize Vault client: %s", err)
+			return fmt.Errorf("Failed to initialize Vault client for %s: %s", cfg.Name, err)
 		}
-		useragent.SetHeaders(f.client)
+		f.clients[cfg.Name] = client
+		useragent.SetHeaders(client)
 	}
 
 	// Connect to vault and parse its information
-	status, err := f.client.Sys().SealStatus()
+	status, err := client.Sys().SealStatus()
 	if err != nil {
-
 		// Print a message indicating that Vault is not available anymore
-		if f.lastState == vaultAvailable {
-			f.logger.Info("Vault is unavailable")
+		if lastState, ok := f.lastStates[cfg.Name]; !ok || lastState == vaultAvailable {
+			f.logger.Info("Vault is unavailable", "cluster", cfg.Name)
```

Comment:
Alternatively, we could create a method-specific logger with the label always applied so we don't forget to add the cluster:

```go
logger := f.logger.With("cluster", cfg.Name)
```